src/libstd/sys/unix/fs.rs

```diff
@@ -44,6 +48,83 @@ pub struct File(FileDesc);
 #[derive(Clone)]
 pub struct FileAttr {
     stat: stat64,
+    #[cfg(target_os = "linux")]
+    statx_extra_fields: Option<StatxExtraFields>,
+}
+
+#[derive(Clone)]
+struct StatxExtraFields {
+    // This is needed to check if btime is supported by the filesystem.
+    stx_mask: u32,
+    stx_btime: libc::statx_timestamp,
+}
+
+// We prefer `statx` on Linux if available, which contains file creation time.
+// Default `stat64` contains no creation time.
+#[cfg(target_os = "linux")]
+unsafe fn try_statx(
+    fd: c_int,
+    path: *const libc::c_char,
+    flags: i32,
+    mask: u32,
+) -> Option<io::Result<FileAttr>> {
+    use crate::sync::atomic::{AtomicBool, Ordering};
+
+    // Linux kernel prior to 4.11 or glibc prior to glibc 2.28 don't support `statx`
+    // We store the availability in a global to avoid unnecessary syscalls
+    static HAS_STATX: AtomicBool = AtomicBool::new(true);
+    syscall! {
+        fn statx(
+            fd: c_int,
+            pathname: *const libc::c_char,
+            flags: c_int,
+            mask: libc::c_uint,
+            statxbuf: *mut statx
+        ) -> c_int
+    }
+
+    if !HAS_STATX.load(Ordering::Relaxed) {
+        return None;
+    }
+
+    let mut buf: statx = mem::zeroed();
+    let ret = cvt(statx(fd, path, flags, mask, &mut buf));
+    match ret {
+        Err(err) => match err.raw_os_error() {
+            Some(libc::ENOSYS) => {
+                HAS_STATX.store(false, Ordering::Relaxed);
+                return None;
+            }
+            _ => return Some(Err(err)),
+        }
+        Ok(_) => {
+            // We cannot fill `stat64` exhaustively because of private padding fields.
+            let mut stat: stat64 = mem::zeroed();
+            stat.st_dev = makedev(buf.stx_dev_major, buf.stx_dev_minor);
+            stat.st_ino = buf.stx_ino;
+            stat.st_nlink = buf.stx_nlink as u64;
+            stat.st_mode = buf.stx_mode as u32;
+            stat.st_uid = buf.stx_uid;
+            stat.st_gid = buf.stx_gid;
+            stat.st_rdev = makedev(buf.stx_rdev_major, buf.stx_rdev_minor);
+            stat.st_size = buf.stx_size as i64;
+            stat.st_blksize = buf.stx_blksize as i64;
+            stat.st_blocks = buf.stx_blocks as i64;
+            stat.st_atime = buf.stx_atime.tv_sec;
+            stat.st_atime_nsec = buf.stx_atime.tv_nsec as i64;
+            stat.st_mtime = buf.stx_mtime.tv_sec;
+            stat.st_mtime_nsec = buf.stx_mtime.tv_nsec as i64;
+            stat.st_ctime = buf.stx_ctime.tv_sec;
+            stat.st_ctime_nsec = buf.stx_ctime.tv_nsec as i64;
+
+            let extra = StatxExtraFields {
+                stx_mask: buf.stx_mask,
```

Comment:
This `stx_mask` field, according to the man page, looks to indicate whether fields are actually available or not. This is used for the creation time, but should this also be taken into account for all the fields above as well? (presumably for them we'd return an error if they aren't available?)

Reply:
So we need to check `buf.statx_mask & STATX_BASIC_STATS == STATX_BASIC_STATS`?

But what would be returned by `stat64` for the case of missing basic fields?
I haven't seen anything about it in the man page of `stat`, including the `ERRORS` section.

For [man page of `statx`](http://man7.org/linux/man-pages/man2/statx.2.html),
> If a filesystem does not support a field or if it has an unrepre‐
> sentable value (for instance, a file with an exotic type), then the
> mask bit corresponding to that field will be cleared in stx_mask even
> if the user asked for it **and a dummy value will be filled in for com‐**
> **patibility purposes if one is available** (e.g., a dummy UID and GID
> may be specified to mount under some circumstances).

Is this the same behavior as `stat` which never return a "partial result"?

BTW: I tested to `statx` a file in mounted FAT32 filesystem which has no permission support. The result has all `STATX_BASIC_STATS` set in `stx_mode`, and the file mode, uid and gid are just what specified in `mount`. And `stat64` also returns the dummy result without any error.

Reply:
Ah ok, this sounds reasonable to me, thanks for the clarification!